Fix crash on empty text file import

mastermakrela opened the pull request:
### What does this PR do?

<!-- **Please explain what your changes do**, example: -->

<!--

This adds a new flag --bail to bun test. When set, it will stop running tests after the first failure. This is useful for CI environments where you want to fail fast.

-->

Fixes https://github.com/oven-sh/bun/issues/19164

![SCR-20250421-sqps](https://github.com/user-attachments/assets/cdfdb80d-5b11-4fa5-9d05-10069dc1bf01)

And possibly addresses regression from 

https://github.com/oven-sh/bun/commit/d1eba784b9d902ac6043c1b76e64382d43fbc261#diff-18858af190099c2a71c98ab1f31358cc47563cf58f2333def11449014677be7fL2752

where the if check changed, to a situation where the second argument was redundant,
i.e., in `...source.contents.len == 0 and source.contents.len < 33...` if first part is true, second also has to be true.


- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

<!-- **For code changes, please include automated tests**. Feel free to uncomment the line below -->

The change is rather small, but if needed I can write some tests.
(I actually found it when testing other PR, so the tests are there)

<!-- I wrote automated tests -->

<!-- If JavaScript/TypeScript modules or builtins changed:

- [ ] I included a test for the new code, or existing tests cover it
- [ ] I ran my tests locally and they pass (`bun-debug test test-file-name.test`)

-->

If Zig files changed:

<!-- 
- [ ] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [ ] I included a test for the new code, or an existing test covers it
- [ ] JSValue used outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
- [ ] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test-file-name.test`)
-->

- I'm new to zig and as far as I can tell I did not touch any memory.
- I'm not sure how one would test those changes, but If someone points me to an example, I'll gladly implement the tests :) (I couldn't file the tests for zig)

<!-- If new methods, getters, or setters were added to a publicly exposed class:

- [ ] I added TypeScript types for the new methods, getters, or setters
-->

<!-- If dependencies in tests changed:

- [ ] I made sure that specific versions of dependencies are used instead of ranged or tagged versions
-->

<!-- If a new builtin ESM/CJS module was added:

- [ ] I updated Aliases in `module_loader.zig` to include the new module
- [ ] I added a test that imports the module
- [ ] I added a test that require() the module
-->

pfgithub commented:
It would be nice to have a test that importing an empty file returns `""`. If there's no test, it could break in the future if the code is changed.

mastermakrela commented:
I've implemented the tests - maybe even too exhaustively - but as I've noticed the inconsistencies with empty files between the transpiler and builder I wanted to be sure everything is ok. 😅

<details>
<summary> Before / after for `test/js/bun/resolve/import-empty.test.js` </summary>

```zsh
❯ bun test import-empty
bun test v1.2.11-canary.24 (27a5712e)

test/js/bun/resolve/import-empty.test.js:
 9 | 
10 | // MARK: - text like
11 | 
12 | it("importing empty text file returns empty string", async () => {
13 |   const empty_file_text = (await import("./empty-file", { with: { type: "text" } })).default;
14 |   expect(empty_file_text).toEqual("");
                               ^
error: expect(received).toEqual(expected)

Expected: ""
Received: undefined

      at <anonymous> (/Users/krzysztof/Developer/mastermakrela/bun/test/js/bun/resolve/import-empty.test.js:14:27)
✗ importing empty text file returns empty string [0.81ms]
✓ importing empty file with type file returns it path [0.14ms]
✓ importing empty css file returns its path [0.05ms]
✓ importing empty html file returns HTMLBundle with its path [0.06ms]
✓ importing empty js like file returns empty module [1.96ms]
✓ importing empty json file throws JSON Parse error [0.24ms]
✓ importing empty jsonc/toml file returns module with empty object as default export [0.15ms]
✓ importing empty file returns module with path as default export [0.13ms]
✓ importing empty sqlite files returns database object [5.19ms]

 8 pass
 1 fail
 22 expect() calls
Ran 9 tests across 1 files. [65.00ms]

     ~/Developer/mastermakrela/bun     fix-text-import-bug ?11 ················································································ 22:43:11   
❯ BUN_DEBUG_QUIET_LOGS=1 build/debug/bun-debug test import-empty
bun test v1.2.11 (f3395152)

test/js/bun/resolve/import-empty.test.js:
✓ importing empty text file returns empty string [4.56ms]
✓ importing empty file with type file returns it path [1.37ms]
✓ importing empty css file returns its path [0.75ms]
✓ importing empty html file returns HTMLBundle with its path [1.70ms]
✓ importing empty js like file returns empty module [7.20ms]
✓ importing empty json file throws JSON Parse error [4.68ms]
✓ importing empty jsonc/toml file returns module with empty object as default export [1.99ms]
✓ importing empty file returns module with path as default export [1.77ms]
✓ importing empty sqlite files returns database object [10.06ms]

 9 pass
 0 fail
 22 expect() calls
Ran 9 tests across 1 files. [618.00ms]
```

</details>


<details>
<summary> Before / after for `test/bundler/bundler_loader.test.ts` </summary>

```zsh
❯ bun test bundler_loader                                         
bun test v1.2.11-canary.24 (27a5712e)

test/bundler/bundler_loader.test.ts:
✓ bundler > bun loader > bun/loader-text-file [39.14ms]
✓ bundler > bun loader > bun/loader-json-file [24.86ms]
✓ bundler > bun loader > bun/loader-toml-file [22.32ms]
✓ bundler > bun loader > bun/loader-text-file [25.92ms]
✓ bundler > node loader > bun/loader-text-file [25.09ms]
✓ bundler > node loader > bun/loader-json-file [27.15ms]
✓ bundler > node loader > bun/loader-toml-file [28.30ms]
✓ bundler > node loader > bun/loader-text-file [26.91ms]
runtime failed file: /private/var/folders/c1/_k2x2fmn5hj7vg1ckc_7dyw00000gn/T/bun-build-tests/bun-1nkwHq/bun/loader-empty-text-file/out.js
stdout output:
{}
---
expected stdout:
""
---
1609 |               console.log(`---`);
1610 |               console.log(`expected ${name}:`);
1611 |               console.log(expected);
1612 |               console.log(`---`);
1613 |             }
1614 |             expect(result).toBe(expected);
                                  ^
error: expect(received).toBe(expected)

Expected: "\"\""
Received: "{}"

      at <anonymous> (/Users/krzysztof/Developer/mastermakrela/bun/test/bundler/expectBundled.ts:1614:28)
      at <anonymous> (/Users/krzysztof/Developer/mastermakrela/bun/test/bundler/expectBundled.ts:561:11)
✗ bundler > handles empty files > bun/loader-empty-text-file [33.57ms]
✓ bundler > bun/loader-text-file [27.46ms]
✓ bundler > bun/wasm-is-copied-to-outdir [38.63ms]
✓ bundler > bun/loader-text-file-#3449 [26.03ms]
✓ bundler > bun/loader-copy-file-entry-point-with-onLoad-wasm [8.40ms]
✓ bundler > bun/loader-copy-file-entry-point-with-onLoad-json [4.80ms]
✓ bundler > bun/loader-copy-file-entry-point-with-onLoad-file [3.61ms]
✓ bundler > bun/loader-copy-file-entry-point-with-onLoad-text [2.74ms]
✓ bundler > bun/loader-copy-file-entry-point-wasm [12.66ms]
✓ bundler > bun/loader-copy-file-entry-point-json [10.24ms]
✓ bundler > bun/loader-copy-file-entry-point-file [13.67ms]
✓ bundler > bun/loader-copy-file-entry-point-text [11.13ms]

 19 pass
 1 fail
 16 expect() calls
Ran 20 tests across 1 files. [588.00ms]

     ~/Developer/mastermakrela/bun     fix-text-import-bug +3 !1 ?10 ·········································································· 23:22:22   
❯ BUN_DEBUG_QUIET_LOGS=1 build/debug/bun-debug test bundler_loader
bun test v1.2.11 (f3395152)

test/bundler/bundler_loader.test.ts:
✓ bundler > bun loader > bun/loader-text-file [350.10ms]
✓ bundler > bun loader > bun/loader-json-file [266.31ms]
✓ bundler > bun loader > bun/loader-toml-file [243.37ms]
✓ bundler > bun loader > bun/loader-text-file [248.40ms]
✓ bundler > node loader > bun/loader-text-file [251.79ms]
✓ bundler > node loader > bun/loader-json-file [248.46ms]
✓ bundler > node loader > bun/loader-toml-file [240.21ms]
✓ bundler > node loader > bun/loader-text-file [242.44ms]
✓ bundler > handles empty files > bun/loader-empty-text-file [240.06ms]
✓ bundler > bun/loader-text-file [235.10ms]
✓ bundler > bun/wasm-is-copied-to-outdir [261.46ms]
✓ bundler > bun/loader-text-file-#3449 [243.27ms]
✓ bundler > bun/loader-copy-file-entry-point-with-onLoad-wasm [34.14ms]
✓ bundler > bun/loader-copy-file-entry-point-with-onLoad-json [19.94ms]
✓ bundler > bun/loader-copy-file-entry-point-with-onLoad-file [17.76ms]
✓ bundler > bun/loader-copy-file-entry-point-with-onLoad-text [17.67ms]
✓ bundler > bun/loader-copy-file-entry-point-wasm [108.29ms]
✓ bundler > bun/loader-copy-file-entry-point-json [103.08ms]
✓ bundler > bun/loader-copy-file-entry-point-file [102.85ms]
✓ bundler > bun/loader-copy-file-entry-point-text [101.79ms]

 20 pass
 0 fail
 16 expect() calls
Ran 20 tests across 1 files. [4.26s]
```

</details>


---

If it's ok, I'd like to finish the bundler tests for other loaders, to get the output to be consistent between transpile and bundle.
(If it's an overkill for this issue, I can open a new one ;) )

mastermakrela commented:
Another not quite bug, but curious thing I've stumbled upon is the imports cache not taking type into account, e.g.:

```ts
import empty_text from "./empty.txt" with { type: "text" };
console.log(`🦔 ~ empty_text:`, empty_text);
console.log(`🦔 ~ typeof empty_text:`, typeof empty_text);

import empty_text2 from "./empty.txt" with { type: "file" };
console.log(`🦔 ~ empty_text2:`, empty_text2);
console.log(`🦔 ~ typeof empty_text2:`, typeof empty_text2);
```

bundles to

```js
// empty.txt
var empty_default = {};

// test.ts
console.log(`\uD83E\uDD94 ~ empty_text:`, empty_default);
console.log(`\uD83E\uDD94 ~ typeof empty_text:`, typeof empty_default);
console.log(`\uD83E\uDD94 ~ empty_text2:`, empty_default);
console.log(`\uD83E\uDD94 ~ typeof empty_text2:`, typeof empty_default);
```

which makes sense for JS-like imports, but for other types I'd kind of expect to be able to get contents of the same file parsed differently.
I'm not sure if this warrants a separate issue?

RiskyMH commented on test/js/bun/resolve/import-empty.test.js:
```diff
@@ -0,0 +1,101 @@
+
+it("importing empty file with type file returns it path", async () => {
+  const empty_file_text = (await import("./empty-file", { with: { type: "file" } })).default;
+  expect(empty_file_text).toEqual(empty_file_path);
```
https://buildkite.com/bun/bun/builds/15329#annotation-test/js/bun/resolve/import-empty.test.js

windows does paths differently btw

mastermakrela commented:
In the last commit, I've updated the tests and changed how the empty files are transformed.

Instead of hard-coding defaults for each loader
https://github.com/oven-sh/bun/pull/19165/commits/f3395152695b8a1a247f46b7ac13bbebf00f518a#diff-c9bc275f9466e5179bb80454b6445c7041d2a0fb79932dd5de7a5c3196bdbd75R4239-R4242

now the loader defines if it should handle the empty files itself
https://github.com/oven-sh/bun/pull/19165/commits/bf1d9a606def51e245dad7973c629775e9bb30a7#diff-b5f3aea99b4d6bca1f40b379bc33eaf925dcb2816b9e74ad94a7dc638990917bR700-R705

This means less places that need to be changed for new loaders.

---

While at it, I've also fixed the `file` type loader.
Earlier, if the file was empty it didn't copy it and bundler wrote an empty object.
Now it correctly copies the empty file and writes the path to it.
